### PR TITLE
libmpdel: enable use of local UNIX sockets for MPD

### PR DESCRIPTION
**Pull request by kennyballou**
By using `make-network-process` we can tune the parameters for either network connections or local UNIX socket connections.

As an added bonus, connections are very fast.  In my limited testing, `make-network-process` tends to be dramatically faster than the `open-network-stream` function.  Of course, without passing `:family 'ipv4` to `make-network-process`, the two functions exhibit similar performance, so it's not as simple as "one is written in C and the other is written in ELisp."

Deprecate/obsolete the `libmpdel-port` variable in favor of `libmpdel-service` which semantically makes more sense now that libmpdel can connect to either a port or a socket.

`libmpdel-profiles` also work with the new local socket connection behaviour.

Although I don't use `customize`, the new variable (`libmpdel-service`) and profiles (`libmpdel-profiles`) should be accessible via `customize` just the same.

### CODE REVIEW

File: libmpdel.el [hunk 1]
```diff
@@ -49,14 +49,42 @@
...
+
+Supported symbol values include:
+  local -- for a local (i.e. UNIX) address specified by SERVICE.
+  ipv4  -- use IPv4 address family only.
+  ipv6  -- use IPv6 address family only.
+
+See `make-network-process' for more information."
+  :type 'symbol)
+
+(defcustom libmpdel-profiles (list(list "Local server" libmpdel-hostname libmpdel-port libmpdel-family))
```
Comment (DamienCassou) [inline libmpdel.el:78]: you removed a space before `(list` and this seems to confuse github diff). Can you please add it back?

File: libmpdel.el [hunk 2]
```diff
@@ -49,14 +49,42 @@
   "MPD server port to connect to.  Also see `libmpdel-hostname'."
   :type 'integer)
 
-(defcustom libmpdel-profiles (list (list "Local server" libmpdel-hostname libmpdel-port))
-  "List of (HOST . PORT) when using several MPD servers."
+(defcustom libmpdel-service 6600
```
Comment (DamienCassou) [inline libmpdel.el:52]: I think renaming port to service makes sense but I would like it to be in a separate commit or PR to ease understanding.

File: libmpdel.el [hunk 3]
```diff
@@ -49,14 +49,42 @@
...
+  "MPD service to connect to.  Also see `libmpdel-hostname'.
+
+This value should be an integer port number (e.g., 6600), an
+integer string port number (e.g., \"6600\"), or a absolute socket
```
Comment (DamienCassou) [inline libmpdel.el:56]: a absolute => an absolute

File: libmpdel.el [hunk 4]
```diff
@@ -49,14 +49,42 @@
...
+  synomous when using port number values.  However, since socket
+  connection support is being added, prefer the `limpdel-service'
+  name to better align with `make-network-process' and be more
+  semantically meaningful.")
+
+(defcustom libmpdel-family 'ipv4
+  "Address and protocol family for MPD service connection.
+
+Supported symbol values include:
+  local -- for a local (i.e. UNIX) address specified by SERVICE.
```
Comment (DamienCassou) [inline libmpdel.el:71]: SERVICE => `libmpdel-service` ?

File: libmpdel.el [hunk 5]
```diff
@@ -49,14 +49,42 @@
...
+  ipv4  -- use IPv4 address family only.
+  ipv6  -- use IPv6 address family only.
+
+See `make-network-process' for more information."
+  :type 'symbol)
+
+(defcustom libmpdel-profiles (list(list "Local server" libmpdel-hostname libmpdel-port libmpdel-family))
+  "List of (HOST . SERVICE . FAMILY) when using several MPD servers."
```
Comment (DamienCassou) [inline libmpdel.el:79]: I think the `NAME` is missing (old documentation bug) and there should only be one dot:

> List of (NAME HOST SERVICE . FAMILY) when using...

File: libmpdel.el [hunk 6]
```diff
@@ -49,14 +49,42 @@
...
+(defcustom libmpdel-family 'ipv4
+  "Address and protocol family for MPD service connection.
+
+Supported symbol values include:
+  local -- for a local (i.e. UNIX) address specified by SERVICE.
+  ipv4  -- use IPv4 address family only.
+  ipv6  -- use IPv6 address family only.
+
+See `make-network-process' for more information."
+  :type 'symbol)
```
Comment (DamienCassou) [inline libmpdel.el:76]: can you please make the type a little more explicit by listing the allowed values?

Maybe with a choice/const (untested):

```emacs
(choice (const :tag "UNIX socket" 'local)
             (const :tag "IPv4 connection" 'ipv4)
             (const :tag "IPv6 connection" 'ipv6))
```

Here is some documentation:
https://www.gnu.org/software/emacs/manual/html_node/elisp/Composite-Types.html#Composite-Types

Can you please test with `M-x customize-variable RET libmpdel-family RET`?

File: libmpdel.el [hunk 7]
```diff
@@ -49,14 +49,42 @@
...
+(defcustom libmpdel-profiles (list(list "Local server" libmpdel-hostname libmpdel-port libmpdel-family))
+  "List of (HOST . SERVICE . FAMILY) when using several MPD servers."
   :type '(repeat (list
                   :tag "Profile"
-                  :value ("Local server" "localhost" 6600)
+                  :value ("Local server" "localhost" 6600 ipv4)
                   (string :tag "name")
                   (string :tag "host")
-                  (integer :tag "port"))))
+                  (choice (integer :tag "port")
```
Comment (DamienCassou) [inline libmpdel.el:85]: same here. You may reference the other variable's type, maybe with https://www.gnu.org/software/emacs/manual/html_node/elisp/Defining-New-Types.html#Defining-New-Types.

**Review by DamienCassou — COMMENTED** [review]
Awesome work, thank you very much. There was already some changes from Ian Zimmerman that I never pushed to github (now I did) so you might have to fix a few conflicts. Sorry about that.

I suggest a few changes below.

As a rule of thumb, I like small and dedicated commits and PRs. If you can split your work into several smaller works. I will be happy. For example, adding support for int and string port numbers should be in a separate commit or PR.  Same thing for the deprecation of `libmpdel-port`.

File: libmpdel.el (hunk 4 above)
Comment (kennyballou) [reply libmpdel.el:71]: I lifted this comment directly from `make-network-process`.  But yes, it should be tailored for the context.

**Comment by kennyballou** [comment]
Looking over the changes recently added to `master`, I'm not sure what use this PR still has.  There are a few small changes that could be included from this PR as a separate PR, e.g., doc fixes and maybe remove `open-network-stream`.  Otherwise, I think what is in `master` now effectively accomplishes what this PR proposes, specifically, `libmpdel` can connect to the local UNIX socket instead of only via TCP.

**Comment by kennyballou** [comment]
I'm closing this in favor of #3 since a5ef4807b209f2ceb0df3df00a60384cc6ba1718 effectively accomplished what I wanted from this PR, i.e., socket support.

@DamienCassou thank you for the comments and suggestions; they were really helpful!
